File: pyopengl/renderer/shader.py
import OpenGL.GL as ogl
import enum
import typing


class ShaderType(enum.Enum):
    VERTEX_SHADER = 0
    FRAGMENT_SHADER = 1


class ShaderException(Exception):
    pass


class Shader:
    def __init__(self, shader_type: str, shader_code: str) -> None:
        self.shader_type = shader_type
        self.shader_code = shader_code
        self.shader = None

        self.compile()

    def compile(self) -> None:
        if self.shader_type == "vertex":
            self.shader = ogl.glCreateShader(ogl.GL_VERTEX_SHADER)
        elif self.shader_type == "fragment":
            self.shader = ogl.glCreateShader(ogl.GL_FRAGMENT_SHADER)
        else:
            raise ShaderException("not implemented")

        ogl.glShaderSource(self.shader, self.shader_code)
        ogl.glCompileShader(self.shader)

        result = ogl.glGetShaderiv(self.shader, ogl.GL_COMPILE_STATUS)

        if not result:
            print("could not compile shader {}".format(self.shader_type))

    def free(self) -> None:
        ogl.glDeleteShader(self.shader)

    @property
    def id(self):
        return self.shader


class Program:
    def __init__(self) -> None:
        self.program = ogl.glCreateProgram()
        self.shaders: typing.List[Shader] = []

    def add_shader(self, shader: Shader) -> None:
        self.shaders.append(shader)

    def compile(self) -> None:
        for shader in self.shaders:
            ogl.glAttachShader(self.program, shader.id)
        ogl.glLinkProgram(self.program)

    def use(self) -> None:
        ogl.glUseProgram(self.program)

    def unuse(self) -> None:
        ogl.glUseProgram(0)

    def validate(self) -> bool:
        ogl.glValidateProgram(self.program)
        validation = ogl.glGetProgramiv(self.program, ogl.GL_VALIDATE_STATUS)
        if validation == ogl.GL_FALSE:
            why = ogl.glGetProgramInfoLog(self.program)
            print("could not compile program: {0}".format(why))
            return False
        return True

    def get_attribute(self, name: str) -> str:
        attr = ogl.glGetAttribLocation(self.program, name)
        return attr

    def get_uniform(self, name: str) -> str:
        attr = ogl.glGetUniformLocation(self.program, name)
        return attr

    def set_uniform(self, name: str, value) -> None:
        ogl.glUniformMatrix4fv(self.get_uniform(name), 1, False, value)

    def set_attribute(self, name: str) -> None:
        # This call relies on the buffer being bound
        ogl.glEnableVertexAttribArray(self.get_attribute(name))
        ogl.glVertexAttribPointer(
            self.get_attribute(name), 3, ogl.GL_FLOAT, ogl.GL_FLOAT, 0, None
        )

File: cgl/shader.c
#include "shader.h"

#include <OpenGL/gl3.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "util.h"

int init_shaderprogram(ShaderProgram *program) {
    memset(program, 0, sizeof(ShaderProgram));

    program->program_id = glCreateProgram();
    printf("created new GL shader program %d\n", program->program_id);
    return 0;
}

void use_shaderprogram(const ShaderProgram *program) {
    glUseProgram(program->program_id);
    printf("using shader program %d\n", program->program_id);
}

int destroy_shaderprogram(ShaderProgram *program) { return 0; }

void add_shader(ShaderProgram *program, ShaderType type, const char *filepath) {
    unsigned int shader_id;
    const char *shader_data = read_file(filepath);
    GLint shader_result = 0;

    if (!shader_data) {
        printf("could not load shader data\n");
        exit(1);
    }

    switch (type) {
        case VERTEXSHADER:
            printf("creating new vertex shader from %s\n", filepath);
            shader_id = glCreateShader(GL_VERTEX_SHADER);
            break;
        case FRAGSHADER:
            printf("creating new fragment shader from %s\n", filepath);
            shader_id = glCreateShader(GL_FRAGMENT_SHADER);
            break;
    }

    glShaderSource(shader_id, 1, &shader_data, NULL);
    glCompileShader(shader_id);
    glGetShaderiv(shader_id, GL_COMPILE_STATUS, &shader_result);

    if (!shader_result) {
        printf("could not compile shader %s\n", filepath);
        glDeleteShader(shader_id);
        exit(1);
    }

    program->shaders[program->shader_count++].shader_id = shader_id;

    free_file_data((char *)shader_data);
    printf("added new shader to program %d in slot %d\n", program->program_id,
           program->shader_count - 1);
}

void compile_shaderprogram(const ShaderProgram *program) {
    for (int i = 0; i < program->shader_count; i++) {
        glAttachShader(program->program_id, program->shaders[i].shader_id);
    }
    glLinkProgram(program->program_id);
    printf("compiled shader program %d\n", program->program_id);
}

void set_attribute(int index, int size) {
    glEnableVertexAttribArray(index);
    glVertexAttribPointer(index, size, GL_FLOAT, GL_FALSE, 0, NULL);
    printf("set attribute for index %d\n", index);
}

void set_uniform(int index, float *value) {
    glUniformMatrix4fv(index, 1, GL_FALSE, value);
}

int get_uniform(const ShaderProgram *program, const char *name) {
    return glGetUniformLocation(program->program_id, name);
}